wdl-lint/src/rules/nonmatching_output.rs

```diff
@@ -0,0 +1,349 @@
+                        self.name.as_deref().expect("should have a name"),
+                        self.ty.expect("should have a type"),
+                    ));
+                } else {
+                    check_matching(state, self);
+                }
+
+                self.name = None;
+                self.current_meta_outputs_span = None;
+                self.current_meta_span = None;
+                self.current_output_span = None;
+                self.output_keys.clear();
+                self.meta_outputs_keys.clear();
+            }
+        }
+    }
+
+    fn task_definition(
+        &mut self,
+        state: &mut Self::State,
+        reason: VisitReason,
+        task: &TaskDefinition,
+    ) {
+        match reason {
+            VisitReason::Enter => {
+                self.name = Some(task.name().as_str().to_string());
+                self.ty = Some("task");
+            }
+            VisitReason::Exit => {
+                if self.current_meta_span.is_some()
+                    && self.current_meta_outputs_span.is_none()
+                    && !self.output_keys.is_empty()
+                {
+                    state.add(missing_outputs_in_meta(
+                        self.current_meta_span.expect("should have a `meta` span"),
+                        self.name.as_deref().expect("should have a name"),
+                        self.ty.expect("should have a type"),
+                    ));
+                } else {
+                    check_matching(state, self);
+                }
+
+                self.current_meta_outputs_span = None;
+                self.current_meta_span = None;
+                self.current_output_span = None;
+                self.output_keys.clear();
+                self.meta_outputs_keys.clear();
+            }
+        }
+    }
+
+    fn metadata_section(
+        &mut self,
+        _state: &mut Self::State,
+        reason: VisitReason,
+        section: &MetadataSection,
+    ) {
+        match reason {
+            VisitReason::Enter => {
+                self.current_meta_span = Some(section.syntax().text_range().to_span());
+            }
+            VisitReason::Exit => {}
+        }
+    }
+
+    fn output_section(
+        &mut self,
+        _state: &mut Self::State,
+        reason: VisitReason,
+        section: &OutputSection,
+    ) {
+        if reason == VisitReason::Enter {
+            self.current_output_span = Some(section.syntax().text_range().to_span());
+        }
+    }
+
+    fn bound_decl(
+        &mut self,
+        _state: &mut Self::State,
+        reason: VisitReason,
+        decl: &wdl_ast::v1::BoundDecl,
+    ) {
+        if reason == VisitReason::Enter {
+            if let Some(output) = &self.current_output_span {
+                let decl_span = decl.syntax().text_range().to_span();
+                if decl_span.start() > output.start() && decl_span.end() < output.end() {
```

Comment:
That's one way to check if the decl is within the output 😄.

Based on my last comment, you might could check if `self.current_output_span` is `None` if you took that approach.